lib/Directives/Toctree.php

```diff
@@ -38,6 +41,8 @@ public function process(
             return;
         }
 
+        assert($node instanceof CodeNode, 'Toctree should be passed a CodeNode only');
```

Comment:
This is enforced (in practice) by how the system works with `wantCode` below. If we were passed some other type of node, it would be unexpected. It allows us to also type-hint `CodeNode` instead of `Node` in `ToctreeBuilder`, which is good, because we can then get more information (like "options" such as `reversed`).

Reply:
I believe this is a bug rather a feature, or a very wrong name of "CodeNode" :) (I think there is some wrong priority in how `DocumentParser` determines what the next block needs to be)

Reply:
What should we do in the next major release? Make the type declaration line 35 stricter? Should we also add an `@param` annotation right now so that static analyzers can detect wrong calls?

Reply:
I'm afraid I don't have any time this week to dig deeper into this. The current parser has 3 very similar states/concepts: Directive, CodeNode and Block. I think we should do some cleanup there. I had a look last week and concluded this:

"CodeNode" is meant to be the indented text following `::` (called ["literal block" in reStructedText](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#literal-blocks)).
```
+------------------------------+
| paragraph                    |
| (ends with "::")             |
+------------------------------+
   +---------------------------+
   | indented literal block    |
   +---------------------------+
```

[Directives](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#directives) have a type, options and optionally a directive body (which, as the previous line ends with `::`, is considered a "CodeNode" by this library):
```
+-------+-------------------------------+
| ".. " | directive type "::" directive |
+-------+ block                         |
        |                               |
        +-------------------------------+
```

And the concept of "block" in the `DocumentParser` doesn't appear to have any related concept in reStructuredText.

Reply:
> I believe this is a bug rather a feature, or a very wrong name of "CodeNode"

The more I look at it, the more I agree that this is a bug, more than a "bad name".

> "CodeNode" is meant to be the indented text following ::

Hmm, that may be true. It seems like:

A) Many directives are returning `true` from `wantsCode()` that should NOT. This is causing a `CodeNode` to be passed to them. `TocTree` is a perfect example. We could work to reverse those - return `false` from those - and instead allow them to work with some generic `Node` object that would be passed to them instead.

B) I'm... not entirely sure... why we need `wantsCode()` at all. Any directive should just be passed... some generic Node. Then it works its magic to create its specific node type (like `TocNode`). If you're right that the `CodeNode` is only meant to represent a literal block... that's just something that `DocumentParser` should handle... and it would not be seen as a directive.

FYI - I'm going to reverse some of these changes.